File: panda-ij-plugin/src/eu/nyerel/panda/ijplugin/runner/PandaSettings.java
package eu.nyerel.panda.ijplugin.runner;

import com.intellij.ide.util.PropertiesComponent;
import com.intellij.openapi.project.Project;
import com.intellij.ui.classFilter.ClassFilter;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Rastislav Papp (rastislav.papp@gmail.com)
 */
public enum PandaSettings {

	INSTANCE;

	private static final String PROPERTY_PANDA_MONITORED_CLASSES = "panda.monitored.classes";

	public ClassFilter[] getMonitoredClasses(Project project) {
		PropertiesComponent properties = PropertiesComponent.getInstance(project);
		String[] classesProperty = properties.getValues(PROPERTY_PANDA_MONITORED_CLASSES);
		if (classesProperty != null) {
			return convertToMonitoredClasses(classesProperty);
		} else {
			return null;
		}
	}

	public String getMonitoredClassesAsString(Project project) {
		String[] monitoredClasses = convertToStringArray(getMonitoredClasses(project));
		StringBuilder sb = new StringBuilder();
		if (monitoredClasses != null) {
			for (int i = 0; i < monitoredClasses.length; i++) {
				sb.append(monitoredClasses[i]);
				if (i < monitoredClasses.length - 1) {
					sb.append(",");
				}
			}
		}
		return sb.toString();
	}

	public void setMonitoredClasses(Project project, ClassFilter[] monitoredClasses) {
		PropertiesComponent properties = PropertiesComponent.getInstance(project);
		properties.setValues(PandaSettings.PROPERTY_PANDA_MONITORED_CLASSES, convertToStringArray(monitoredClasses));
	}

	@NotNull
	private ClassFilter[] convertToMonitoredClasses(String[] monitoredClassPatterns) {
		List<ClassFilter> classFilterList = new ArrayList<ClassFilter>();
		for (String monitoredClassPattern : monitoredClassPatterns) {
			ClassFilter classFilter = new ClassFilter(monitoredClassPattern);
			classFilter.setEnabled(true);
			classFilterList.add(classFilter);
		}
		return classFilterList.toArray(new ClassFilter[classFilterList.size()]);
	}

	private String[] convertToStringArray(ClassFilter[] monitoredClasses) {
		if (monitoredClasses == null) {
			return null;
		} else {
			List<String> list = new ArrayList<String>();
			for (ClassFilter monitoredClass : monitoredClasses) {
				if (monitoredClass.isEnabled()) {
					list.add(monitoredClass.getPattern());
				}
			}
			return list.toArray(new String[list.size()]);
		}
	}

}

File: panda-ij-plugin/src/eu/nyerel/panda/ijplugin/runner/PandaSettings.kt
package eu.nyerel.panda.ijplugin.runner

import com.intellij.ide.util.PropertiesComponent
import com.intellij.openapi.project.Project
import com.intellij.ui.classFilter.ClassFilter

import java.util.ArrayList

private const val PROPERTY_PANDA_MONITORED_CLASSES = "panda.monitored.classes"
private const val PROPERTY_PANDA_EXCLUDED_CLASSES = "panda.excluded.classes"

/**
 * @author Rastislav Papp (rastislav.papp@gmail.com)
 */
object PandaSettings {

    var isAggregateCallTree = true
    var isCallTreeStructured = true
    var isSortByDuration = true
    var isShowPackageName = false
    var hideProxyClasses = true

    fun getMonitoredClasses(project: Project): Array<ClassFilter> {
        return getClasses(project, PROPERTY_PANDA_MONITORED_CLASSES)
    }

    fun getExcludedClasses(project: Project): Array<ClassFilter> {
        return getClasses(project, PROPERTY_PANDA_EXCLUDED_CLASSES)
    }

    private fun getClasses(project: Project, propertyName: String): Array<ClassFilter> {
        val properties = PropertiesComponent.getInstance(project)
        val classesProperty = properties.getValues(propertyName)
        return if (classesProperty != null) {
            convertToMonitoredClasses(classesProperty)
        } else {
            emptyArray()
        }
    }

    fun getMonitoredClassesAsString(project: Project): String {
        val monitoredClasses = convertToStringArray(getMonitoredClasses(project))
        return convertArrayToString(monitoredClasses)
    }

    fun getExcludedClassesAsString(project: Project): String {
        val monitoredClasses = convertToStringArray(getExcludedClasses(project))
        return convertArrayToString(monitoredClasses)
    }

    private fun convertArrayToString(monitoredClasses: Array<String>?): String {
        val sb = StringBuilder()
        if (monitoredClasses != null) {
            for (i in monitoredClasses.indices) {
                sb.append(monitoredClasses[i])
                if (i < monitoredClasses.size - 1) {
                    sb.append(",")
                }
            }
        }
        return sb.toString()
    }

    fun setMonitoredClasses(project: Project, monitoredClasses: Array<ClassFilter>) {
        val properties = PropertiesComponent.getInstance(project)
        properties.setValues(PROPERTY_PANDA_MONITORED_CLASSES, convertToStringArray(monitoredClasses))
    }

    fun setExcludedClasses(project: Project, excludedClasses: Array<ClassFilter>) {
        val properties = PropertiesComponent.getInstance(project)
        properties.setValues(PROPERTY_PANDA_EXCLUDED_CLASSES, convertToStringArray(excludedClasses))
    }

    private fun convertToMonitoredClasses(monitoredClassPatterns: Array<String>): Array<ClassFilter> {
        val classFilterList = ArrayList<ClassFilter>()
        for (monitoredClassPattern in monitoredClassPatterns) {
            val classFilter = ClassFilter(monitoredClassPattern)
            classFilter.isEnabled = true
            classFilterList.add(classFilter)
        }
        return classFilterList.toTypedArray()
    }

    private fun convertToStringArray(monitoredClasses: Array<ClassFilter>?): Array<String>? {
        return if (monitoredClasses == null) {
            null
        } else {
            val list = ArrayList<String>()
            for (monitoredClass in monitoredClasses) {
                if (monitoredClass.isEnabled) {
                    list.add(monitoredClass.pattern)
                }
            }
            list.toTypedArray()
        }
    }

}
